### ISSUE TITLE
Happywhale - Whale and Dolphin Identification

### ISSUE THREAD

**Issue by ztsv-av** [ztsv-av/spellbook#25]
New competition ! ! ! Yay

This competition is about predicting an 'individual_id' from a given image of a dolphin or a whale. Images usually contain a fin and a back of an animal.

For each image in the test set, we may predict up to 5 'individual_id' labels. There are individuals in the test set that are not seen in the training data; these should be predicted as 'new_individual'.

Submission file explanation:
- In the first column 'image' we should submit the image id, the reality it is just the filenames of the images (e.g 00021adfb725ed.jpg, 000562241d384d.jpg, etc.) and in the other column 'predictions' top 5 predictions in the right order (from most confident to the less confident, descending order if we would have the probabilities) for the given image, predictions' format is 5 ids (e.g 00021adfb725ed, 000562241d384d, etc.), which are divided by space, e.g ("00021adfb725ed 000562241d384d").

Metrics used in this competition is Mean Average Precision @ 5 (MAP@5). I will code this metric myself and used it during training.

Now, a few notices and ideas that I got from reading all of the discussion posts:
1. Do not drop grayscale images - convert them to RGB
2. Remember covid19 competition? There were around 4 classes and we wrote prediction score for all of them (there was a discussion post that stated that this increased MAP@5 metric score). I think we should try to do the same here - write up to 5 most probable individual_id scores for every image. Also, I thought of an another idea. "There are individuals in the test set that are not seen in the training data; these should be predicted as 'new_individual'" - what if we write 'new_individual' for every image at the last place of a prediction row? We can try that aswell.
3. We can try first creating a model/ensemble to predict whether a given animal is a dolphin or a whale. Then, we train another model that predicts 'individual_id' based on whether it is a whale or a dolphin. This will decrease the number of classes for the model to predict drastically. However, competition host said: "One potential problem here is that some individuals may be easier to detect directly rather than labeling to species first. For example, fin, sei and bryde's whales all have very similar dorsal fins, where we can expect more variation between individuals than between species". I am not sure, but we can try that, I think we will have a plenty of time.
4. Decide on whether to apply augmentation techniques or not.
5. Decide which loss to use? We can try to use Focal Loss or Contrastive Loss.

**Comment by ztsv-av** [comment]
- Check focal loss, for some reason the last data in batch has Nan loss value (only last part of the batch). Or it is the loss for the whole batch, need to check
- data loading takes a lot of time, probably because of the size of the metadata. Need to create one hot while loading data

**Comment by ztsv-av** [comment]
Created function that creates class one-hot vector while creating a label, not loading it from the metadata. It reduced data loading time significantly.

**Comment by ztsv-av** [comment]
MODEL / LB SCORE
InceptionV3 / 0.17 (384x384)
EffNetB5 /  0.65 (768x768)
EffNetB6 / 0.72 (768x768)

Trained around 20 epochs

**Comment by ztsv-av** [comment]
Folds and models ensemble all increase score.
Image size also greatly increase score.

**Comment by powerswing** [comment]
TODO after analysis:

- [ ] add **ArcMarginProduct** to model
- [ ] unfreeze all model layers during train
- [ ] add **Dense()** layer in the end
- [ ] add loss and metric as per code below
- [ ] probably use a lower initial learning rate (divide current by 10 or 100?)

```py
metrics = [tf.keras.metrics.SparseCategoricalAccuracy(),tf.keras.metrics.SparseTopKCategoricalAccuracy(k=5)]
loss = [tf.keras.losses.SparseCategoricalCrossentropy()],
```
